text/0000-const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,673 @@
+        Foo(self.0 + other.0)
+    }
+}
+#[derive(Debug)]
+struct Bar;
+impl Add for Bar {
+    fn add(self, other: Self) -> Self {
+        println!("hello from the otter side: {:?}", other);
+        self
+    }
+}
+impl Neg for Bar {
+    fn neg(self) -> Self {
+        self
+    }
+}
+```
+
+allows calling `Foo(Bar) + Foo(Bar)` even though that is most definitely not const,
+because `Bar` only has an `impl Add for Bar`
+and not an `impl const Add for Bar`. Expressed in some sort of effect system syntax (neither
+effect syntax nor effect semantics are proposed by this RFC, the following is just for demonstration
+purposes):
+
+```rust
+impl<c: constness, T: const(c) Add> const(c) Add for Foo<T> {
+    const(c) fn add(self, other: Self) -> Self {
+        Foo(self.0 + other.0)
+    }
+}
+```
+
+In this scheme on can see that if the `c` parameter is set to `const`, the `T` parameter requires a
+`const Add` bound, and creates a `const Add` impl for `Foo<T>` which then has a `const fn add`
+method. On the other hand, if `c` is `?const`, we get a regular impl without any constness anywhere.
+For regular impls one can still pass a `T` which has a `const Add` impl, but that won't
+cause any constness for `Foo<T>`.
+
+This goes in hand with the current scheme for const functions, which may also be called
+at runtime with runtime arguments, but are checked for soundness as if they were called in
+a const context. E.g. the following function may be called as
+`add(Bar, Bar)` at runtime.
+
+```rust
+const fn add<T: Neg, U: Add<T>>(a: T, b: U) -> T {
+    -a + b
+}
+```
+
+Using the same effect syntax from above:
+
+```rust
+<c: constness> const(c) fn add<T: const(c) Neg, U: const(c) Add<T>>(a: T, b: U) -> T {
+    -a + b
+}
+```
+
+Here the value of `c` decides both whether the `add` function is `const` and whether its parameter
+`T` has a `const Add` impl. Since both use the same `constness` variable, `T` is guaranteed to have
+a `const Add` iff `add` is `const`.
+
+This feature could have been added in the future in a backwards compatible manner, but without it
+the use of `const` impls is very restricted for the generic types of the standard library due to
+backwards compatibility.
+Changing an impl to only allow generic types which have a `const` impl for their bounds would break
+situations like the one described above.
+
+## `?const` opt out
+
+There is often desire to add bounds to a `const` function's generic arguments, without wanting to
+call any of the methods on those generic bounds. Prominent examples are `new` functions:
+
+```rust
+struct Foo<T: Trait>(T);
+const fn new<T: Trait>(t: T) -> Foo<T> {
+    Foo(t)
+}
+```
+
+Unfortunately, with the given syntax in this RFC, one can now only call the `new` function in a const
+context if `T` has
+an `impl const Trait for T { ... }`. Thus an opt-out similar to `?Sized` can be used:
+
+```rust
+struct Foo<T: Trait>(T);
+const fn new<T: ?const Trait>(t: T) -> Foo<T> {
+    Foo(t)
+}
+```
```

Comment:
I think it would be a good idea to relate this to the `c: constness` notation from the previous section -- to explain how the `?const` affects the "desugaring" into the more expressive calculus.